frame/support/src/storage/counted_map.rs

```diff
@@ -0,0 +1,314 @@
+use codec::{FullCodec, Decode, EncodeLike, Encode};
+use crate::{
+	storage::{
+		StorageMap, StorageValue,
+		StorageAppend, StorageTryAppend, StorageDecodeLength, StoragePrefixedMap,
+		generator::{StorageMap as StorageMapT, StorageValue as StorageValueT},
+	},
+};
+use sp_std::prelude::*;
+use sp_runtime::traits::Saturating;
+
+/// A wrapper around a `StorageMap` and a `StorageValue<u32>` to keep track of how many items are in
+/// a map, without needing to iterate all the values.
+///
+/// This storage item has additional storage read and write overhead when manipulating values
+/// compared to a regular storage map.
+///
+/// For functions where we only add or remove a value, a single storage read is needed to check if
+/// that value already exists. For mutate functions, two storage reads are used to check if the
+/// value existed before and after the mutation.
+///
+/// Whenever the counter needs to be updated, an additional read and write occurs to update that
+/// counter.
+pub struct CountedStorageMap<Key, Value, Map, Counter>(
+	core::marker::PhantomData<(Key, Value, Map, Counter)>
+);
+
+impl<Key, Value, Map, Counter>
+	StorageMapT<Key, Value>
+	for CountedStorageMap<Key, Value, Map, Counter>
+where
+	Key: FullCodec,
+	Value: FullCodec,
+	Map: StorageMapT<Key, Value>,
+	Counter: StorageValueT<u32>
+{
+	type Query = Map::Query;
+	type Hasher = Map::Hasher;
+	fn module_prefix() -> &'static [u8] {
+		Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Map::storage_prefix()
+	}
+	fn from_optional_value_to_query(v: Option<Value>) -> Self::Query {
+		Map::from_optional_value_to_query(v)
+	}
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value> {
+		Map::from_query_to_optional_value(v)
+	}
+}
+
+impl<Key, Value, Map, Counter>
+	StoragePrefixedMap<Value> for
```

Comment:
storage prefix map has some defaults method that allow to remove values, and you don't override them here.
I think you should override them, but at the same time I think it is a bit dangerous to continue having default methods if some implementation needs to override them.
In case somebody add a new method which break the CountedStorageMap assumption then he must update the CountedStorageMap.
I would personnally prefer moving the method of StoragePrefixedMap into the trait StorageMap itself, and don't implement StoragePrefixedMap.